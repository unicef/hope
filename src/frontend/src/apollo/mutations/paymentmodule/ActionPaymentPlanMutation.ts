import { gql } from '@apollo/client';

export const ACTION_PAYMENT_PLAN = gql`
  mutation ActionPP($input: ActionPaymentPlanInput!) {
    actionPaymentPlanMutation(input: $input) {
      paymentPlan {
        id
        status
      }
    }
  }
`;
