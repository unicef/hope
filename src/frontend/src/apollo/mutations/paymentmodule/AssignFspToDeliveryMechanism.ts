import { gql } from '@apollo/client';

export const ASSIGN_FSP_TO_DELIVERY_MECHANISM = gql`
  mutation AssignFspToDeliveryMech($input: AssignFspToDeliveryMechanismInput!) {
    assignFspToDeliveryMechanism(input: $input) {
      paymentPlan {
        id
        deliveryMechanisms {
          id
          name
          fsp {
            id
            name
          }
        }
      }
    }
  }
`;
