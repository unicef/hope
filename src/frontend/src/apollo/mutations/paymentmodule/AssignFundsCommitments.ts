import { gql } from '@apollo/client';

export const ASSIGN_FUNDS_COMMITMENTS = gql`
  mutation AssignFundsCommitmentsPaymentPlan(
    $paymentPlanId: ID!
    $fundCommitmentItemsIds: [String!]!
  ) {
    assignFundsCommitments(
      paymentPlanId: $paymentPlanId
      fundCommitmentItemsIds: $fundCommitmentItemsIds
    ) {
      paymentPlan {
        id
        availableFundsCommitments {
          fundsCommitmentNumber
          fundsCommitmentItems {
            paymentPlan {
              id
              name
            }
            fundsCommitmentItem
            recSerialNumber
          }
        }
      }
    }
  }
`;
