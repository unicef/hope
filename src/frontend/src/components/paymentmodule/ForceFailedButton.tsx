import { Box, Button, DialogContent, DialogTitle } from '@mui/material';
import * as React from 'react';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Dialog } from '@containers/dialogs/Dialog';
import { DialogActions } from '@containers/dialogs/DialogActions';
import { DialogContainer } from '@containers/dialogs/DialogContainer';
import { DialogFooter } from '@containers/dialogs/DialogFooter';
import { DialogTitleWrapper } from '@containers/dialogs/DialogTitleWrapper';
import { useSnackbar } from '@hooks/useSnackBar';
import { useMarkPayAsFailedMutation } from '@generated/graphql';

export interface ForceFailedButtonProps {
  paymentId: string;
  disabled?: boolean;
}
export function ForceFailedButton({
  paymentId,
  disabled = false,
}: ForceFailedButtonProps): React.ReactElement {
  const { t } = useTranslation();
  const [isOpenModal, setOpenModal] = useState(false);
  const { showMessage } = useSnackbar();
  const [mutate, { loading }] = useMarkPayAsFailedMutation();

  const submit = async (): Promise<void> => {
    try {
      await mutate({
        variables: {
          paymentId,
        },
      });
      setOpenModal(false);
      showMessage(t('Payment has been marked as failed.'));
    } catch (e) {
      e.graphQLErrors.map((x) => showMessage(x.message));
    }
  };

  return (
    <Box>
      <Box p={2}>
        <Button
          color="primary"
          variant="contained"
          onClick={() => setOpenModal(true)}
          data-cy="button-mark-as-failed"
          disabled={disabled}
        >
          {t('Mark as failed')}
        </Button>
      </Box>
      <Dialog
        open={isOpenModal}
        onClose={() => setOpenModal(false)}
        scroll="paper"
        aria-labelledby="form-dialog-title"
        maxWidth="md"
      >
        <DialogTitleWrapper>
          <DialogTitle>{t('Mark as failed')}</DialogTitle>
        </DialogTitleWrapper>
        <DialogContent>
          <DialogContainer>
            <Box p={5}>
              {t('Are you sure you would like to mark payment as failed?')}
            </Box>
          </DialogContainer>
        </DialogContent>
        <DialogFooter>
          <DialogActions>
            <Button onClick={() => setOpenModal(false)}>{t('CANCEL')}</Button>
            <Button
              type="submit"
              color="primary"
              variant="contained"
              onClick={() => submit()}
              data-cy="button-submit"
              disabled={loading}
            >
              {t('Mark as failed')}
            </Button>
          </DialogActions>
        </DialogFooter>
      </Dialog>
    </Box>
  );
}
