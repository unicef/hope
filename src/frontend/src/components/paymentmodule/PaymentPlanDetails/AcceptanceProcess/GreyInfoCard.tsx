import { Box } from '@mui/material';
import * as React from 'react';
import styled from 'styled-components';
import { UniversalMoment } from '@core/UniversalMoment';
import { PaymentPlanQuery } from '@generated/graphql';
import { MessageDialog } from './MessageDialog';

const GreyText = styled.div`
  color: #9e9e9e;
`;

const GreyTitle = styled.div`
  color: #7c8990;
  text-transform: uppercase;
  font-size: 12px;
`;

const IconPlaceholder = styled.div`
  width: 16px;
  height: 16px;
`;

const GreyBox = styled(Box)`
  background-color: #f4f5f6;
`;

interface GreyInfoCardProps {
  topMessage: string;
  topDate: string;
  approvals: PaymentPlanQuery['paymentPlan']['approvalProcess']['edges'][number]['node']['actions']['approval'];
}

export function GreyInfoCard({
  topMessage,
  topDate,
  approvals,
}: GreyInfoCardProps): React.ReactElement {
  const mappedApprovals = approvals.map((action) => {
    const { info, createdAt, comment, createdBy } = action;
    return (
      info && (
        <Box alignItems="center" display="flex" key={createdAt}>
          {info}
          <Box ml={1}>
            <GreyText>
              on <UniversalMoment>{createdAt}</UniversalMoment>
            </GreyText>
          </Box>
          <Box p={1} ml={1}>
            {comment ? (
              <MessageDialog
                comment={comment}
                author={createdBy}
                date={createdAt}
              />
            ) : (
              <IconPlaceholder />
            )}
          </Box>
        </Box>
      )
    );
  });

  return (
    <Box display="flex" flexDirection="column">
      <Box p={3}>
        <GreyTitle>
          {topMessage} on
          <UniversalMoment>{topDate}</UniversalMoment>
        </GreyTitle>
      </Box>
      <GreyBox
        display="flex"
        flexDirection="column"
        alignItems="center"
        ml={3}
        mr={3}
        p={3}
      >
        {mappedApprovals}
      </GreyBox>
    </Box>
  );
}
