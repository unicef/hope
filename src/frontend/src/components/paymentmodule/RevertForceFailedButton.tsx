import { Box, Button, DialogContent, DialogTitle } from '@mui/material';
import * as React from 'react';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Field, Form, Formik } from 'formik';
import * as Yup from 'yup';
import CalendarTodayRoundedIcon from '@mui/icons-material/CalendarTodayRounded';
import { Dialog } from '@containers/dialogs/Dialog';
import { DialogActions } from '@containers/dialogs/DialogActions';
import { DialogFooter } from '@containers/dialogs/DialogFooter';
import { DialogTitleWrapper } from '@containers/dialogs/DialogTitleWrapper';
import { useSnackbar } from '@hooks/useSnackBar';
import { FormikTextField } from '@shared/Formik/FormikTextField';
import { FormikDateField } from '@shared/Formik/FormikDateField';
import { useRevertMarkPayAsFailedMutation } from '@generated/graphql';

export interface RevertForceFailedButtonProps {
  paymentId: string;
  disabled?: boolean;
}
export function RevertForceFailedButton({
  paymentId,
  disabled = false,
}: RevertForceFailedButtonProps): React.ReactElement {
  const { t } = useTranslation();
  const [isOpenModal, setOpenModal] = useState(false);
  const { showMessage } = useSnackbar();
  const [mutate, { loading }] = useRevertMarkPayAsFailedMutation();

  const validationSchema = Yup.object().shape({
    deliveredQuantity: Yup.number()
      .min(0)
      .max(99999999, t('Number is too big')),
    deliveryDate: Yup.date().required(t('Delivery date is required')),
  });

  const submit = async (values, { resetForm }): Promise<void> => {
    try {
      await mutate({
        variables: {
          paymentId,
          deliveredQuantity: values.deliveredQuantity,
          deliveryDate: values.deliveryDate,
        },
      });
      setOpenModal(false);
      showMessage(t('Force failed has been reverted.'));
      resetForm();
    } catch (e) {
      e.graphQLErrors.map((x) => showMessage(x.message));
    }
  };

  return (
    <Box>
      <Box p={2}>
        <Button
          color="primary"
          variant="contained"
          onClick={() => setOpenModal(true)}
          data-cy="button-revert-mark-as-failed"
          disabled={disabled}
        >
          {t('Revert mark as failed')}
        </Button>
      </Box>
      <Formik
        initialValues={{
          deliveredQuantity: 0,
          deliveryDate: '',
        }}
        validationSchema={validationSchema}
        onSubmit={submit}
      >
        {({ submitForm, resetForm }) => (
          <Dialog
            open={isOpenModal}
            onClose={() => setOpenModal(false)}
            scroll="paper"
            aria-labelledby="form-dialog-title"
            maxWidth="md"
          >
            <DialogTitleWrapper>
              <DialogTitle>{t('Revert mark as failed')}</DialogTitle>
            </DialogTitleWrapper>
            <DialogContent>
              <Form>
                <Field
                  name="deliveredQuantity"
                  label={t('Delivered Quantity')}
                  type="number"
                  required
                  fullWidth
                  variant="outlined"
                  component={FormikTextField}
                />
                <Field
                  name="deliveryDate"
                  label={t('Delivery Date')}
                  component={FormikDateField}
                  required
                  fullWidth
                  decoratorEnd={<CalendarTodayRoundedIcon color="disabled" />}
                />
              </Form>
            </DialogContent>
            <DialogFooter>
              <DialogActions>
                <Button
                  onClick={() => {
                    setOpenModal(false);
                    resetForm();
                  }}
                >
                  {t('CANCEL')}
                </Button>
                <Button
                  type="submit"
                  color="primary"
                  variant="contained"
                  onClick={submitForm}
                  data-cy="button-submit"
                  disabled={loading}
                >
                  {t('Revert mark as failed')}
                </Button>
              </DialogActions>
            </DialogFooter>
          </Dialog>
        )}
      </Formik>
    </Box>
  );
}
