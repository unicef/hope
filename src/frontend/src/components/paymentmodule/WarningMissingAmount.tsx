import { Box } from '@mui/material';
import WarningIcon from '@mui/icons-material/Warning';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import { ReactElement } from 'react';

const WarnIcon = styled(WarningIcon)`
  color: ${({ theme }) => theme.hctPalette.orange};
`;

interface WarningMissingAmountProps {
  amount: number;
  currency: string;
}

export function WarningMissingAmount({
  amount,
  currency,
}: WarningMissingAmountProps): ReactElement {
  const { t } = useTranslation();

  return (
    <Box display="flex" alignItems="center">
      <WarnIcon />
      {t('Missing')} {amount} {currency}
    </Box>
  );
}
