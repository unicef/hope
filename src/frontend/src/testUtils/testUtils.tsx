import { ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, RenderOptions } from '@testing-library/react';
import { Formik } from 'formik';
import noop from 'lodash/noop';
import { TestProviders } from './testProviders';
import { ProgramStatus } from '@generated/graphql';
import { MockLink } from '@apollo/client/testing';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) =>
  render(
    <>
      <BrowserRouter>
        <Formik initialValues={{}} onSubmit={noop}>
          {ui}
        </Formik>
      </BrowserRouter>
    </>,
    {
      wrapper: TestProviders,
      ...options,
    },
  );

export * from '@testing-library/react';
export { customRender as render };

let m_w = 123456789;
let m_z = 987654321;
const mask = 0xffffffff;

// Takes any integer
export const seed = (i) => {
  m_w = (123456789 + i) & mask;
  m_z = (987654321 - i) & mask;
};

// Returns number between 0 (inclusive) and 1.0 (exclusive),
// just like Math.random().
export const random = () => {
  m_z = (36969 * (m_z & 65535) + (m_z >> 16)) & mask;
  m_w = (18000 * (m_w & 65535) + (m_w >> 16)) & mask;
  let result = ((m_z << 16) + (m_w & 65535)) >>> 0;
  result /= 4294967296;
  return result;
};

export class ApolloLoadingLink extends MockLink {
  constructor() {
    super([]);
  }

  request() {
    return null;
  }
}

export const fakeContextProgram = {
  selectedProgram: {
    id: '1',
    name: 'someName',
    status: ProgramStatus.Active,
    pduFields: null,
    dataCollectingType: {
      id: '1',
      householdFiltersAvailable: true,
      individualFiltersAvailable: true,
      label: 'data collecting type',
      code: '123',
      type: 'full',
      children: null,
    },
  },
  setSelectedProgram: () => {},
  isActiveProgram: true,
  isSocialDctType: false,
  isStandardDctType: true,
  programHasPdu: false,
};
