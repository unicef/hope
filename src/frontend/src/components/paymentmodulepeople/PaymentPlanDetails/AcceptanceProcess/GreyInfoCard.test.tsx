import * as React from 'react';
import { fakeApolloPaymentPlan } from '../../../../../fixtures/paymentmodule/fakeApolloPaymentPlan';
import { render } from '../../../../testUtils/testUtils';
import { GreyInfoCard } from './GreyInfoCard';

describe('components/paymentmodule/PaymentPlanDetails/AcceptanceProcess/GreyInfoCard', () => {
  it('should render', () => {
    const { container } = render(
      <GreyInfoCard
        topMessage="Test top message"
        topDate="2022-01-01"
        approvals={
          fakeApolloPaymentPlan.approvalProcess.edges[0].node.actions.approval
        }
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
