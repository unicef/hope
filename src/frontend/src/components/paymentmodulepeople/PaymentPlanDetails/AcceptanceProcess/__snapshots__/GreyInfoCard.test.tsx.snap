// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/paymentmodule/PaymentPlanDetails/AcceptanceProcess/GreyInfoCard should render 1`] = `
<div>
  <div
    class="MuiBox-root css-j7qwjs"
  >
    <div
      class="MuiBox-root css-38zrbw"
    >
      <div
        class="sc-ixGGxD foWIaj"
      >
        Test top message
         on
        <time
          datetime="1640995200000"
        >
          1 Jan 2022
        </time>
      </div>
    </div>
    <div
      class="sc-dstKZu chYUKh MuiBox-root css-e6xxd1"
    >
      <div
        class="MuiBox-root css-fv3lde"
      >
        Approved by Root Rootkowski
        <div
          class="MuiBox-root css-15fzge"
        >
          <div
            class="sc-kLhKbu fsBovP"
          >
            on 
            <time
              datetime="1661843741787"
            >
              30 Aug 2022
            </time>
          </div>
        </div>
        <div
          class="MuiBox-root css-1el3gal"
        >
          <div
            class="sc-khLCKb kDroYd"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
